casper/src/main/scala/coop/rchain/casper/BlockCreator.scala

```diff
@@ -0,0 +1,233 @@
+                                .toList
+    } yield (validDeploys -- deploysInCurrentChain.flatten).toSeq
+
+  private def notExpiredDeploy(earliestBlockNumber: Long, d: DeployData): Boolean =
+    d.validAfterBlockNumber > earliestBlockNumber
+
+  private def notFutureDeploy(currentBlockNumber: Long, d: DeployData): Boolean =
+    d.validAfterBlockNumber < currentBlockNumber
+
+  /*
+   * We ensure that only the justifications given in the block are those
+   * which are bonded validators in the chosen parent. This is safe because
+   * any latest message not from a bonded validator will not change the
+   * final fork-choice.
+   */
+  private def computeJustifications[F[_]: Monad](
+      dag: BlockDagRepresentation[F],
+      parents: Seq[BlockMessage]
+  ): F[Seq[Justification]] = {
+    val bondedValidators = bonds(parents.head).map(_.validator).toSet
+    dag.latestMessages.map { latestMessages =>
+      toJustification(latestMessages)
+        .filter(j => bondedValidators.contains(j.validator))
+    }
+  }
+
+  private def processDeploysAndCreateBlock[F[_]: Sync: Log: BlockStore](
+      dag: BlockDagRepresentation[F],
+      runtimeManager: RuntimeManager[F],
+      parents: Seq[BlockMessage],
+      deploys: Seq[DeployData],
+      justifications: Seq[Justification],
+      maxBlockNumber: Long,
+      shardId: String,
+      version: Long,
+      now: Long
+  ): F[CreateBlockStatus] =
+    InterpreterUtil
+      .computeDeploysCheckpoint[F](
+        parents,
+        deploys,
+        dag,
+        runtimeManager,
+        Some(now)
+      )
+      .flatMap {
+        case Left(ex) =>
+          Log[F]
+            .error(
+              s"Critical error encountered while processing deploys: ${ex.getMessage}"
+            )
+            .map(_ => CreateBlockStatus.internalDeployError(ex))
+
+        case Right((preStateHash, postStateHash, processedDeploys)) =>
+          val (internalErrors, persistableDeploys) =
+            processedDeploys.partition(_.status.isInternalError)
+          internalErrors.toList
+            .traverse {
+              case InternalProcessedDeploy(deploy, _, _, InternalErrors(errors)) =>
+                val errorsMessage = errors.map(_.getMessage).mkString("\n")
+                Log[F].error(
+                  s"Internal error encountered while processing deploy ${PrettyPrinter
+                    .buildString(deploy)}: $errorsMessage"
+                )
+              case _ => ().pure[F]
+            } >>
+            runtimeManager
+              .computeBonds(postStateHash)
+              .map { newBonds =>
+                createBlock(
+                  now,
+                  parents,
+                  justifications,
+                  maxBlockNumber,
+                  preStateHash,
+                  postStateHash,
+                  persistableDeploys,
+                  newBonds,
+                  shardId,
+                  version
+                )
+              }
+      }
```

Comment:
the function processDeploysAndCreateBlock is pretty difficult to read. If you split it a little bit it should help the reviewer